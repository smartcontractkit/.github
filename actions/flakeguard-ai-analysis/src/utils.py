import json
from pathlib import Path

import dspy
import pandas as pd
import requests


def get_repo_file(
    file_path: str,
    username: str = "smartcontractkit",
    repo: str = "chainlink",
    branch: str = "develop",
) -> str:
    raw_url = (
        f"https://raw.githubusercontent.com/{username}/{repo}/{branch}/{file_path}"
    )

    response = requests.get(raw_url)
    response.raise_for_status()
    return response.text


def init() -> None:
    llm = dspy.LM("openai/gpt-4o-mini")
    dspy.configure(lm=llm)


def read_results_file(file: Path) -> pd.DataFrame:
    data = json.loads(file.read_text())
    results = data.get("results", [])
    return pd.DataFrame(results)
