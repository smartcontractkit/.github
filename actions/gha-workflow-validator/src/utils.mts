import { GithubFiles } from "./github.mjs";

export interface ParsedFile {
  filename: string;
  sha: string;
  addedLines: FileAddition[];
}

export interface FileAddition {
  lineNumber: number;
  content: string;
  actionReference?: ActionReference;
}

export interface ActionReference {
  owner: string;
  repo: string;
  repoPath: string;
  ref: string;
  comment?: string;
  line: string;
}

export function parseAllAdditions(files: GithubFiles): ParsedFile[] {
  if (!files) return [];

  return files?.map(entry => {
    const { filename, sha, patch } = entry;
    const addedLines = patch ? parsePatchAdditions(patch) : [];
    return { filename, sha, addedLines };
  });
}

function parsePatchAdditions(patch: string): FileAddition[]  {
  const lineChanges = patch?.split('\n') || [];

  const additions: FileAddition[] = [];

  let currentLineInFile = 0;
  for (const line of lineChanges) {

    if (line.startsWith('@@')) {
      // @@ denotes a git hunk header
      // https://mattscodecave.com/posts/howto-reading-git-diffs-and-staging-hunks.html
      // example line: @@ -16,10 +16,10 @@ jobs:
      //   - "-16,10": "-" denotes source file, "16,10" means the hunk starts at line 16 and output contains 10 lines
      //   - "+16,10": "+" denotes destination file, "16,10" means the same as above
      const [ _, source, dest, __ ] = line.split(' ');

      if (!dest.startsWith("+")) {
        throw new Error("Invalid git hunk format");
      }

      const [ destLine, ___ ] = dest.substring(1).split(',');
      currentLineInFile = parseInt(destLine, 10);
      continue;
    } else if (line.startsWith('+')) {
      const currentLine = line.substring(1);
      const actionReference = extractActionReference(currentLine);
      additions.push({ content: currentLine, lineNumber: currentLineInFile, actionReference });
    } else if (line.startsWith('-')) {
      // ignore deletions
      continue;
    }
    currentLineInFile++;
  }

  return additions;
}

export function extractActionReference(line: string): ActionReference | undefined {
  // example line:
  //       - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
  const trimSubString = "uses:"
  const usesIndex = line.indexOf(trimSubString);

  if (usesIndex === -1) {
    // Not an action reference
    return;
  }

  // trim past the "uses:" substring to get "<owner>/<repo><optional path>@<ref> # <optional comment>""
  const trimmedLine = line.substring(line.indexOf(trimSubString) + trimSubString.length).trim();

  const [ actionIdentifier, ...comment ] = trimmedLine.split("#");
  const [ identifier, gitRef ] = actionIdentifier.trim().split("@");
  const [ owner, repo, ...path] = identifier.split("/");
  const repoPath = ((path.length) > 0 ? "/" : "") + path.join("/");

  return { owner, repo, repoPath, ref: gitRef, comment: comment.join().trim(), line, };
}