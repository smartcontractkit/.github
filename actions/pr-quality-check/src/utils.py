"""General utility functions."""

import os
import re
import yaml
from typing import Any, Dict, List

from .glob_match import matches_glob_pattern


def env_bool(name: str, default: bool = False) -> bool:
    """Parse environment variable as boolean (only 'true' is truthy)."""
    v = os.getenv(name)
    if v is None:
        return default
    return str(v).lower() == "true"


def load_yaml_config(path: str) -> Dict[str, Any]:
    """Load YAML configuration file."""
    with open(path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f) or {}


def load_prompt(name: str) -> str:
    """Load prompt template from markdown file."""
    prompt_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "prompts", f"{name}.md")
    with open(prompt_path, "r", encoding="utf-8") as f:
        return f.read().strip()


def to_list(x) -> List[str]:
    """Convert value to list of strings."""
    if x is None:
        return []
    if isinstance(x, list):
        return [str(i) for i in x]
    return [str(x)]


def match_files(files: List[str], patterns: List[str], excludes: List[str]) -> List[str]:
    """Filter files by inclusion/exclusion glob patterns."""
    if not patterns:
        return []
    out: List[str] = []
    for f in files:
        # exclusion first
        excluded = False
        for ex in (excludes or []):
            if ex and matches_glob_pattern(f, ex):
                excluded = True
                break
        if excluded:
            continue
        # inclusion if any pattern matches
        if any(matches_glob_pattern(f, p) for p in patterns if p):
            out.append(f)
    return out


def sanitize_patch(patch: str) -> str:
    """Sanitize git patch by removing code after @@ hunk headers."""
    if not patch:
        return patch
    lines = []
    for ln in patch.splitlines():
        m = re.match(r"(@@[^@]*@@).*$", ln)
        if m:
            lines.append(m.group(1))
        else:
            lines.append(ln)
    return "\n".join(lines)


def build_prompt(template: str, file_path: str, diff: str, rules_yaml: str, matched_patterns_desc: str, required_context_block: str) -> str:
    """Build analysis prompt by substituting placeholders."""
    return template.format(
        file_classification=f"Matched patterns: {matched_patterns_desc}",
        quality_rules=rules_yaml,
        context_files=required_context_block or "No additional context files needed.",
        file_path=file_path,
        diff_output=diff or "(no patch available from GitHub API)"
    )
