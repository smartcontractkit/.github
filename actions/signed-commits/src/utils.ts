import unified from "unified";
import type { Node, Parent } from "unist";
import remarkParse from "remark-parse";
import remarkStringify from "remark-stringify";
// @ts-ignore
import mdastToString from "mdast-util-to-string";
import { getPackages, Package } from "@manypkg/get-packages";
import { exec, getExecOutput } from "@actions/exec";

export const BumpLevels = {
  dep: 0,
  patch: 1,
  minor: 2,
  major: 3,
} as const;

export async function getVersionsByDirectory(cwd: string) {
  let { packages } = await getPackages(cwd);
  return new Map(packages.map((x) => [x.dir, x.packageJson.version]));
}

export async function getChangedPackages(
  cwd: string,
  previousVersions: Map<string, string>,
) {
  let { packages } = await getPackages(cwd);
  let changedPackages = new Set<Package>();

  for (let pkg of packages) {
    const previousVersion = previousVersions.get(pkg.dir);
    if (previousVersion !== pkg.packageJson.version) {
      changedPackages.add(pkg);
    }
  }

  return [...changedPackages];
}

export function getChangelogEntry(changelog: string, version: string) {
  let ast = unified().use(remarkParse).parse(changelog);

  let highestLevel: number = BumpLevels.dep;
  if (!isParentNode(ast)) {
    throw new Error("ast is not a parent node");
  }

  let nodes = ast.children;
  let headingStartInfo:
    | {
        index: number;
        depth: number;
      }
    | undefined;
  let endIndex: number | undefined;

  for (let i = 0; i < nodes.length; i++) {
    let node = nodes[i];

    if (node.type === "heading") {
      if (!nodeHasDepthProperty(node)) {
        throw new Error("node is missing depth property");
      }
      let stringified: string = mdastToString(node);
      let match = stringified.toLowerCase().match(/(major|minor|patch)/);
      if (match !== null) {
        let level = BumpLevels[match[0] as "major" | "minor" | "patch"];
        highestLevel = Math.max(level, highestLevel);
      }
      if (headingStartInfo === undefined && stringified === version) {
        headingStartInfo = {
          index: i,
          depth: node.depth,
        };
        continue;
      }
      if (
        endIndex === undefined &&
        headingStartInfo !== undefined &&
        headingStartInfo.depth === node.depth
      ) {
        endIndex = i;
        break;
      }
    }
  }
  if (headingStartInfo) {
    ast.children = (ast.children as any).slice(
      headingStartInfo.index + 1,
      endIndex,
    );
  }
  return {
    content: unified().use(remarkStringify).stringify(ast),
    highestLevel: highestLevel,
  };
}

export function sortTheThings(
  a: { private: boolean; highestLevel: number },
  b: { private: boolean; highestLevel: number },
) {
  if (a.private === b.private) {
    return b.highestLevel - a.highestLevel;
  }
  if (a.private) {
    return 1;
  }
  return -1;
}

export async function execWithOutput(
  cmd: Parameters<typeof exec>[0],
  args: Parameters<typeof exec>[1],
  opts?: { notrim?: boolean; cwd?: string },
) {
  let { exitCode, stdout, stderr } = await getExecOutput(cmd, args, opts);

  if (!opts?.notrim) {
    stdout = stdout.trim();
    stderr = stderr.trim();
  }

  if (exitCode !== 0) {
    throw Error(stderr);
  }

  return stdout;
}

function isParentNode(node: Node<any>): node is Parent {
  return "children" in node;
}

function nodeHasDepthProperty(
  node: Node<any>,
): node is Node & { depth: number } {
  return "depth" in node;
}
