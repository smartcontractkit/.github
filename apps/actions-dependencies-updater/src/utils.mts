import { existsSync } from "node:fs";
import { readdir } from "node:fs/promises";
import { join } from "node:path";
import * as semver from "semver";

import * as log from "./logger.mjs";
import { Action, WorkflowByName } from "./workflows.mjs";

/**
 * Get an environment variable or throw an error if it's not set
 * @param envVariableName The name of the environment variable
 * @returns The value of the environment variable
 */
export function getEnvironmentVariableOrExit(envVariableName: string): string {
  const variable = process.env[envVariableName];
  if (!variable) {
    log.error(`Mandatory environment variable not set: ${envVariableName}`);
    process.exit(1);
  }
  return variable;
}

export function validateRepositoryOrExit(repoDir: string) {
  if (!repoDir) {
    log.error("No repository directory provided.");
    process.exit(1);
  }

  if (!existsSync(repoDir)) {
    log.error(`Directory does not exist: ${repoDir}`);
    process.exit(1);
  }

  const gitDir = join(repoDir, ".git");
  if (!existsSync(gitDir)) {
    log.error(`Directory is not a git repository: ${repoDir}`);
    process.exit(1);
  }

  const workflowDir = join(repoDir, ".github", "workflows");
  if (!existsSync(workflowDir)) {
    log.info(
      `No workflows directory found: ${workflowDir} - nothing to check/update`,
    );
    process.exit(0);
  }
}

/**
 * From a list of tags, "guess" the latest version.
 * @param tags A list of tags, without the prefixed with "refs/tags/"
 * @returns The latest version
 */
export function guessLatestVersion(
  tags: string[],
  repo?: string,
  repoPath?: string,
) {
  let versions: VersionIdentifier[] = tags
    .map((tag) => parseTagToVersion(tag))
    .filter((v) => !!v) as VersionIdentifier[];

  // support for the .github monorepo
  if (repo === ".github" && repoPath) {
    let actionName = repoPath.split("/").pop();
    if (actionName === "signed-commits")
      actionName = "changesets-signed-commits"; // hacky fix
    log.debug(
      `Guessing latest version for monorepo action: ${actionName} (${repo})`,
    );
    if (actionName) {
      versions = versions.filter((v) => v.prefix === actionName);
    }
  }

  // Sort the versions ascending
  versions.sort((a, b) => semver.compare(a.version, b.version));
  return versions[versions.length - 1];
}

type VersionIdentifier = NonNullable<ReturnType<typeof parseTagToVersion>>;

/**
 * Parse a tag to a version object
 * @param tag The tag to parse
 * @returns The version object
 */
function parseTagToVersion(tag: string) {
  if (tag.startsWith("untagged-")) {
    return;
  }

  const originalTag = tag;
  let prefix = "";

  if (tag.includes("@")) {
    const parts = tag.split("@");
    tag = parts[1];
    prefix = parts[0];
  }

  const coerced = semver.coerce(tag);
  if (!coerced) {
    log.debug(`Failed to parse version from tag: ${tag}`);
    return;
  }

  return {
    major: coerced.major,
    minor: coerced.minor,
    patch: coerced.patch,
    version: coerced.version,
    prefix: prefix,
    tag: originalTag,
  };
}

/**
 * Lists all yml or yaml files within a directory
 * @param directory the directory/path
 * @returns A list of all yml/yaml files contained in that directory
 */
export async function listAllYamlFiles(directory: string) {
  if (!existsSync(directory)) {
    log.warn("Directory does not exist: ", directory);
    return [];
  }

  const files = await readdir(directory).then((files) =>
    files
      .map((f) => join(directory, f))
      .filter((f) => f.includes(".yml") || f.includes(".yaml")),
  );

  return files;
}

export async function getActionYamlPath(directory: string) {
  const yamlFiles = (await listAllYamlFiles(directory)).filter(
    (file) => file.endsWith("action.yml") || file.endsWith("action.yaml"),
  );

  if (yamlFiles.length === 0) {
    log.warn(
      "Found invalid reference to local action yaml. No action file found: ",
      directory,
    );
    return;
  }

  return yamlFiles[0];
}

export function compileDeprecatedPaths(workflowsByName: WorkflowByName) {
  const allDeprecatedPaths: string[] = [];

  for (const workflow of Object.values(workflowsByName)) {
    for (const job of workflow.jobs) {
      const deprecatedPaths = job.dependencies
        .filter(
          (dependency) =>
            dependency.type === "node12" || dependency.type === "node16",
        )
        .map(createPathStrings)
        .flat();
      allDeprecatedPaths.push(...deprecatedPaths);
    }
  }

  const uniquePaths = Array.from(new Set(allDeprecatedPaths)).sort();
  return uniquePaths;
}

function createPathStrings(action: Action): string[] {
  return action.referencePaths.map((path) =>
    [...path, action.identifier, action.type].join(" -> "),
  );
}

/**
 * Given an action identifier, check if it uses a sha reference. This is to ensure that the parsed
 * action was from an immutable reference.
 * @param identifier the action identifier
 * @returns true if the identifier is a sha reference
 */
export function isShaRefIdentifier(identifier: string) {
  const sha1Regex = /^[0-9a-f]{40}$/;
  const sha256Regex = /^[0-9a-f]{256}$/;
  const ref = identifier.split("@")[1];
  return (ref && sha1Regex.test(ref)) || sha256Regex.test(ref);
}
