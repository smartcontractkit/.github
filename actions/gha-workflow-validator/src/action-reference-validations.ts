import { ActionReference, FileAddition, ParsedFile } from './utils.js';
import { Octokit, getActionFileFromGithub } from './github.js';
import * as core from "@actions/core";

export interface ValidationResult {
  filename: string;
  sha: string;
  lineValidations: LineValidationResult[];
}

interface LineValidationResult {
  line: FileAddition;
  validationErrors: ValidationError[];
}

type ValidationError = { message: string; };

export async function validateActionReferenceChanges(octokit: Octokit, changes: ParsedFile[]): Promise<ValidationResult[]> {
  core.debug(`Validating action reference changes, on ${changes.length} changes`)

  const resultsPromise = changes.map(async (change) => {
    const lineValidationPromises  = change.addedLines.filter(addedLine => addedLine.actionReference).map(async (line) => {
      const validationErrors = (line.actionReference) ? await validateLine(octokit, line.actionReference) : [];
      return { line: line, validationErrors };
    });

    const nonEmptyLineValidations = (await Promise.all(lineValidationPromises)).filter((error) => error.validationErrors.length > 0);

    return {
      filename: change.filename,
      sha: change.sha,
      lineValidations: nonEmptyLineValidations,
    };
  });

  const filteredResults = (await Promise.all(resultsPromise)).filter((result) => result.lineValidations.length > 0);
  core.debug(`Found ${filteredResults.length} files with validation errors`);
  return filteredResults
}

async function validateLine(octokit: Octokit, line: ActionReference): Promise<ValidationError[]> {
  const validationErrors: ValidationError[] = [];

  const shaRefValidation = usesShaRef(line);
  const versionCommentValidation = hasVersionComment(line);
  const node20ActionValidation = await isNode20Action(octokit, line);

  if (shaRefValidation) {
    validationErrors.push({ message: shaRefValidation });
  }
  if (versionCommentValidation) {
    validationErrors.push({ message: versionCommentValidation });
  }
  if (node20ActionValidation) {
    validationErrors.push({ message: node20ActionValidation });
  }

  return validationErrors;
}

function usesShaRef(change: ActionReference) {
  const sha1Regex = /^[0-9a-f]{40}$/;
  if(sha1Regex.test(change.ref)) return;

  return `${change.ref} is not a valid SHA1`;
}

function hasVersionComment(change: ActionReference) {
  if (change.comment) return;

  return `No version comment found`
}

async function isNode20Action(ghClient: Octokit, change: ActionReference) {
  const actionFile = await getActionFileFromGithub(
    ghClient,
    change.owner,
    change.repo,
    change.repoPath,
    change.ref,
  );

  if (!actionFile) {
    core.warning(`No action file found for ${change.owner}/${change.repo}${change.repoPath}@${change.ref}`);
    return;
  }

  const nodeVersionRegex = /^\s+using:\s*"?node(\d{2})"?/gm;
  const matches = nodeVersionRegex.exec(actionFile);
  if (matches && matches[1] !== '20') {
    return `Action is using node${matches[1]}`;
  }

  return;
}