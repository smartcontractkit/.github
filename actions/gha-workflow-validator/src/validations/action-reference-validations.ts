import { mapAndFilterUndefined, ParsedFile, FileLine } from "../utils.js";
import { Octokit, getActionFileFromGithub } from "../github.js";
import * as core from "@actions/core";
import {
  ValidationCheck,
  ValidationMessage,
  FileValidationResult,
  LineValidationResult,
  ValidationType,
} from "./validation-check.js";

const CURRENT_NODE_VERSION = 20;

interface FileLineActionReference extends FileLine {
  actionReference?: ActionReference;
}

interface ActionReference {
  owner: string;
  repo: string;
  repoPath: string;
  ref: string;
  comment?: string;
}

interface ActionReferenceValidationOptions {
  validateNodeVersion: boolean;
}

export class ActionReferenceValidation implements ValidationCheck {
  private options: ActionReferenceValidationOptions;

  constructor(
    readonly octokit: Octokit,
    options?: ActionReferenceValidationOptions,
  ) {
    this.options = options ?? { validateNodeVersion: true };
  }

  async validate(parsedFile: ParsedFile): Promise<FileValidationResult> {
    core.debug(`Validating action references in ${parsedFile.filename}`);
    const { filename } = parsedFile;

    const lineActionRefs = mapAndFilterUndefined(
      parsedFile.lines,
      extractActionReference,
    );

    const lineValidations: LineValidationResult[] =
      await validateActionReferences(
        this.octokit,
        filename,
        this.options,
        lineActionRefs,
      );

    return {
      filename,
      lineValidations,
    };
  }
}

async function validateActionReferences(
  octokit: Octokit,
  filename: string,
  options: ActionReferenceValidationOptions,
  lines: FileLineActionReference[],
): Promise<LineValidationResult[]> {
  const lineValidationResults: LineValidationResult[] = [];

  for (const line of lines) {
    const validationErrors = await validateActionReference(
      octokit,
      options,
      line.actionReference,
    );

    if (validationErrors.length > 0) {
      lineValidationResults.push({
        filename,
        line,
        messages: validationErrors,
      });
    }
  }

  return lineValidationResults;
}

async function validateActionReference(
  octokit: Octokit,
  options: ActionReferenceValidationOptions,
  actionRef: ActionReference | undefined,
): Promise<ValidationMessage[]> {
  if (!actionRef) {
    return [];
  }

  const validationErrors: ValidationMessage[] = [];

  const shaRefValidation = validateShaRef(actionRef);
  const versionCommentValidation = validateVersionCommentExists(actionRef);
  const node20ActionValidation = options.validateNodeVersion
    ? await validateNodeActionVersion(octokit, actionRef)
    : undefined;

  if (shaRefValidation) {
    validationErrors.push(shaRefValidation);
  }
  if (versionCommentValidation) {
    validationErrors.push(versionCommentValidation);
  }
  if (node20ActionValidation) {
    validationErrors.push(node20ActionValidation);
  }

  return validationErrors;
}

function validateShaRef(
  actionReference: ActionReference,
): ValidationMessage | undefined {
  const sha1Regex = /^[0-9a-f]{40}$/;
  if (sha1Regex.test(actionReference.ref)) return;

  const sha256Regex = /^[0-9a-f]{256}$/;
  if (sha256Regex.test(actionReference.ref)) return;

  return {
    message: `${actionReference.ref} is not a valid SHA reference`,
    type: ValidationType.SHA_REF,
    severity: "error",
  };
}

function validateVersionCommentExists(
  actionReference: ActionReference,
): ValidationMessage | undefined {
  if (actionReference.comment) return;

  return {
    message: `No version comment found`,
    type: ValidationType.VERSION_COMMENT,
    severity: "warning",
  };
}

async function validateNodeActionVersion(
  octokit: Octokit,
  actionRef: ActionReference,
): Promise<ValidationMessage | undefined> {
  const actionFile = await getActionFileFromGithub(
    octokit,
    actionRef.owner,
    actionRef.repo,
    actionRef.repoPath,
    actionRef.ref,
  );

  if (!actionFile) {
    core.warning(
      `No action file found for ${actionRef.owner}/${actionRef.repo}${actionRef.repoPath}@${actionRef.ref}`,
    );
    return;
  }

  const nodeVersionRegex = /^\s+using:\s*"?node(\d{2})"?/gm;
  const matches = nodeVersionRegex.exec(actionFile);
  if (matches && matches[1] !== `${CURRENT_NODE_VERSION}`) {
    return {
      message: `Action is using node${matches[1]}`,
      type: ValidationType.NODE_VERSION,
      severity: "warning",
    };
  }

  return;
}

function extractActionReference(
  fileLine: FileLine,
): FileLineActionReference | undefined {
  const actionReference = extractActionReferenceFromLine(fileLine.content);
  if (!actionReference) {
    return;
  }

  return {
    ...fileLine,
    actionReference,
  };
}

// Only exported for use in tests
export function extractActionReferenceFromLine(
  line: string,
): ActionReference | undefined {
  const trimmedLine = line.trim();

  if (trimmedLine.startsWith("#")) {
    // commented line
    return;
  }

  // example line:
  // - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
  const trimSubString = "uses:";
  const usesIndex = trimmedLine.indexOf(trimSubString);

  if (usesIndex === -1) {
    // Not an action reference
    return;
  }

  // trim past the "uses:" substring to get "<owner>/<repo><optional path>@<ref> # <optional comment>""
  const trimmedUses = line
    .substring(line.indexOf(trimSubString) + trimSubString.length)
    .trim();

  if (trimmedUses.startsWith("./")) {
    // Local action reference - do not extract or validate these.
    return;
  }

  const [actionIdentifier, ...comment] = trimmedUses.split("#");
  const [identifier, gitRef] = actionIdentifier.trim().split("@");
  const [owner, repo, ...path] = identifier.split("/");
  const repoPath = (path.length > 0 ? "/" : "") + path.join("/");

  return {
    owner,
    repo,
    repoPath,
    ref: gitRef,
    comment: comment.join().trim(),
  };
}
