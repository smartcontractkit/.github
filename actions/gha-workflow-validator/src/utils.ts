import * as core from "@actions/core";
import { GithubFiles } from "./github.js";
import { ValidationResult } from "./action-reference-validations.js";
import { FIXING_ERRORS, htmlLink } from "./strings.js";

export interface ParsedFile {
  filename: string;
  sha: string;
  addedLines: FileAddition[];
}

export interface FileAddition {
  lineNumber: number;
  content: string;
  actionReference?: ActionReference;
}

export interface ActionReference {
  owner: string;
  repo: string;
  repoPath: string;
  ref: string;
  comment?: string;
  line: string;
}

export function filterForGithubWorkflowChanges(files: GithubFiles): GithubFiles {
  return files?.filter(({ filename }) => {
    return (filename.startsWith('.github/workflows') || filename.startsWith('.github/actions'))
    && (filename.endsWith('.yml') || filename.endsWith('.yaml'))
  })
}

export function parseAllAdditions(files: GithubFiles): ParsedFile[] {
  if (!files) return [];

  return files?.map(entry => {
    const { filename, sha, patch } = entry;
    const addedLines = patch ? parsePatchAdditions(patch) : [];
    return { filename, sha, addedLines };
  });
}

function parsePatchAdditions(patch: string): FileAddition[]  {
  const lineChanges = patch?.split('\n') || [];

  const additions: FileAddition[] = [];

  let currentLineInFile = 0;
  for (const line of lineChanges) {

    if (line.startsWith('@@')) {
      // @@ denotes a git hunk header
      // https://mattscodecave.com/posts/howto-reading-git-diffs-and-staging-hunks.html
      // example line: @@ -16,10 +16,10 @@ jobs:
      //   - "-16,10": "-" denotes source file, "16,10" means the hunk starts at line 16 and output contains 10 lines
      //   - "+16,10": "+" denotes destination file, "16,10" means the same as above
      const [ , , dest, ] = line.split(' ');

      if (!dest.startsWith("+")) {
        throw new Error("Invalid git hunk format");
      }

      const [ destLine, ] = dest.substring(1).split(',');
      currentLineInFile = parseInt(destLine, 10);
      continue;
    } else if (line.startsWith('+')) {
      const currentLine = line.substring(1);
      const actionReference = extractActionReference(currentLine);
      additions.push({ content: currentLine, lineNumber: currentLineInFile, actionReference });
    } else if (line.startsWith('-')) {
      // ignore deletions
      continue;
    }
    currentLineInFile++;
  }

  return additions;
}

function extractActionReference(line: string): ActionReference | undefined {
  // example line:
  //       - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
  const trimSubString = "uses:"
  const usesIndex = line.indexOf(trimSubString);

  if (usesIndex === -1) {
    // Not an action reference
    return;
  }

  // trim past the "uses:" substring to get "<owner>/<repo><optional path>@<ref> # <optional comment>""
  const trimmedLine = line.substring(line.indexOf(trimSubString) + trimSubString.length).trim();

  if (trimmedLine.startsWith("./")) {
    // Local action reference - do not extract or validate these.
    return;
  }

  const [ actionIdentifier, ...comment ] = trimmedLine.split("#");
  const [ identifier, gitRef ] = actionIdentifier.trim().split("@");
  const [ owner, repo, ...path] = identifier.split("/");
  const repoPath = ((path.length) > 0 ? "/" : "") + path.join("/");

  return { owner, repo, repoPath, ref: gitRef, comment: comment.join().trim(), line, };
}

export function logErrors(validationResults: ValidationResult[], annotatePR: boolean = false) {
  for (const fileResults of validationResults) {
    for (const lineResults of fileResults.lineValidations) {
      const message = lineResults.validationErrors.map(error => error.message).join(",");
      core.error(`file: ${fileResults.filename} @ line: ${lineResults.line.lineNumber} - ${message}`);
      if (annotatePR) {
        core.error(message, {
          file: fileResults.filename,
          startLine: lineResults.line.lineNumber,
        });
      }
    }
  }
}

type TableRow = Parameters<typeof core.summary.addTable>[0][0]
type TableCell = TableRow[0]

export async function setSummary(validationResults: ValidationResult[], fileUrlPrefix: string) {
  const headerRow: TableRow = [
      { data: "Filename", header: true },
      { data: "Line Number", header: true },
      { data: "Violations", header: true },
  ];

  const errorRows = validationResults.reduce<TableRow[]>((acc, curr) => {
    const filename = curr.filename;

    const errorCellTuples: TableCell[][]  = curr.lineValidations.map(validationResult => {
      const lineNumberCell: TableCell = { data: htmlLink(`${validationResult.line.lineNumber}`, `${fileUrlPrefix}/${filename}#L${validationResult.line.lineNumber}`) }
      const violationsCell: TableCell = { data: validationResult.validationErrors.map(error => error.message).join(", ") }

      return [ lineNumberCell, violationsCell ];
    })

    if (errorCellTuples.length === 0) {
      return acc;
    }

    // The filename cell to span all the rows for this file
    const filenameCell: TableCell = { data: filename, rowspan: `${errorCellTuples.length}` }
    const firstErrorCellTuple =  errorCellTuples.shift() as TableCell[];
    const firstRowForFile = [ filenameCell, ...firstErrorCellTuple ];

    return [
      ...acc,
      firstRowForFile,
      ...errorCellTuples
    ];

  }, [] as TableRow[]);

  await core.summary
    .addTable([headerRow, ...errorRows])
    .addSeparator()
    .addRaw(FIXING_ERRORS)
    .write();
}

