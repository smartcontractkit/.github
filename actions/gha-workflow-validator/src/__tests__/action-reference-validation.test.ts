import { validateActionReferenceChanges } from '../action-reference-validations.js';
import { getTestOctokit } from './__helpers__/test-utils.js'
import nock from "nock";
import path from "path";
import { ParsedFile } from '../utils.js';

// nock-back provides the recording and playback functionality
const nockBack = nock.back;
// Set the fixture path and nockBack mode
nockBack.fixtures = path.join(__dirname, "__fixtures__");

// Change to 'lockdown' to use existing fixtures
// Valid values = lockdown, record, wild, dryrun, update
nockBack.setMode("record");

if (nockBack.currentMode === "lockdown") {
  nock.disableNetConnect();
}

describe("validateActionReferenceChanges", () => {



  it("should validate no changes", async () => {
    const octokit = getTestOctokit(nockBack.currentMode);
    const result = await validateActionReferenceChanges(octokit, []);
    expect(result).toEqual([]);
  });



  it("should validate no action references", async () => {
    const octokit = getTestOctokit(nockBack.currentMode);
    const noWorkflowChanges: ParsedFile[] = [
      {
        filename: ".github/workflows/test.yml",
        sha: "abc",
        addedLines: [
          { lineNumber: 1, content: "line 1" },
          { lineNumber: 2, content: "line 2" },
        ],
      }
    ];
    const result = await validateActionReferenceChanges(octokit, noWorkflowChanges);
    expect(result).toEqual([]);
  });



  it("should validate single action reference", async () => {
    const { nockDone } = await nockBack("actions-checkout-validation.json");
    const octokit = getTestOctokit(nockBack.currentMode);

    const simpleChanges: ParsedFile[] = [
      {
        filename: ".github/workflows/test.yml",
        sha: "abc",
        addedLines: [
          { lineNumber: 1, content: "      - name: test step"},
          { lineNumber: 2,
            content: "        actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1",
            actionReference: {
              owner: "actions",
              repo: "checkout",
              repoPath: "",
              ref: "b4ffde65f46336ab88eb53be808477a3936bae11",
              comment: "v4.1.1",
              line: "        actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1",
            }
          }
        ],
      }
    ];

    const result = await validateActionReferenceChanges(octokit, simpleChanges);
    expect(result).toEqual([]);
    nockDone();
  });
});
