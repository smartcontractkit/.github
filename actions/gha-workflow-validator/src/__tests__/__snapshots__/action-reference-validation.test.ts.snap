// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionRefValidation > should invalidate action reference (sha-ref / no comment) 1`] = `
[
  {
    "message": "No version comment found",
    "severity": "warning",
    "type": "version-comment",
  },
]
`;

exports[`ActionRefValidation > should invalidate single action reference (all errors) 1`] = `
[
  {
    "message": "v2.11.0 is not a valid SHA reference",
    "severity": "error",
    "type": "sha-ref",
  },
  {
    "message": "No version comment found",
    "severity": "warning",
    "type": "version-comment",
  },
  {
    "message": "Action is using node16",
    "severity": "warning",
    "type": "node-version",
  },
]
`;

exports[`ActionRefValidation > should invalidate single action reference (bad sha ref) 1`] = `
[
  {
    "message": "v3.0.2 is not a valid SHA reference",
    "severity": "error",
    "type": "sha-ref",
  },
]
`;

exports[`ActionRefValidation > should invalidate single action reference (no version comment) 1`] = `
[
  {
    "message": "No version comment found",
    "severity": "warning",
    "type": "version-comment",
  },
]
`;

exports[`ActionRefValidation > should invalidate single action reference (node16) 1`] = `
[
  {
    "message": "Action is using node16",
    "severity": "warning",
    "type": "node-version",
  },
]
`;

exports[`ActionRefValidation > should validate single action reference (node24) 1`] = `[]`;
